File: haskell/src/Shapes/Base/Commons/BaseShape.hs
module Shapes.Base.Commons.BaseShape (BaseShape (..)) where

class BaseShape a where
    dimensions :: a -> [(String, Double)]

    printDescription :: a -> IO ()

    showAvailableDimensionCount :: a -> Int
    showAvailableDimensionCount shape = length (dimensions shape)

    validateShape :: a -> Bool
    validateShape _ = True

File: kotlin/src/main/kotlin/shapes/base/commons/BaseShape.kt
package shapes.base.commons

abstract class BaseShape {
    private val _dimensions: List<Map<String, Double>>

    constructor(dimensions: List<Map<String, Double>>) {
        this._dimensions = dimensions
    }

    val dimensions: List<Map<String, Double>>
        get() = _dimensions

    fun showAvailableDimensionCount(): Int {
        return dimensions.size
    }

    companion object {
        fun validateShape(): Boolean {
            return true
        }
    }

    abstract fun printDescription()
}
